gcc/rust/ast/rust-ast-dump.cc

```diff
@@ -0,0 +1,753 @@
+// Copyright (C) 2020-2022 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-ast-dump.h"
+
+namespace Rust {
+namespace AST {
+
+Indent::Indent () : tabs (0) {}
+
+std::ostream &
+operator<< (std::ostream &stream, const Indent &indent)
+{
+  for (size_t i = 0; i < indent.tabs; i++)
+    stream << '\t';
+
+  return stream;
+}
```

Comment:
I've run a benchmark for appending the tab character using a loop vs a string constructor

```cpp
#include <benchmark/benchmark.h>
#include <ostream>

#define tabs 5e+5

static void BM_AppendUsingStringConstructor(benchmark::State &state) {
  std::ostream stream(NULL);
  for (auto _ : state)
    stream << std::string(tabs, '\t');
}

BENCHMARK(BM_AppendUsingStringConstructor);
//
static void BM_AppendUsingInsertionOperator(benchmark::State &state) {
  std::ostream stream(NULL);
  for (auto _: state)
    for (size_t i = 0; i < tabs; i++)
      stream << '\t';
}
BENCHMARK(BM_AppendUsingInsertionOperator);
//
BENCHMARK_MAIN();
```

Some numbers:

Tabs set to `5e+5`
```stdout
--------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations
--------------------------------------------------------------------------
BM_AppendUsingStringConstructor      11169 ns        10953 ns        66106
BM_AppendUsingInsertionOperator    5920976 ns      5913260 ns          117
```

Here is a one-liner too

```bash
for tabs in 1e+3, 1e+5, 1e+10, 1e+20, 1e+50; do echo "Tabs set to ${tabs}\n" && sed -i "s/TABS/${tabs}/" benchmark.cc && g++ main.cpp -std=c++11 -isystem benchmark/include -Lbenchmark/build/src -lbenchmark -lpthread -o benchmark && ./benchmark; done
```

--- 

The number of iterations is massively higher for string constructors, but takes much less time however.

Reply:
You may want to check how other parts of GCC are doing this same kind of tasks (gimple dump would look really close to this for example, I guess). I know that when looking at dumping the HIR in s-expr, looking at the RTL dumper was useful :)

Reply:
Thats a really good point about RTL that will be good for @ndrwnaguib to look at.

Reply:
thank you for sharing that, I will definitely take a look on the RTL dumper.